### PR TITLE
Fix inconsistent number formatting between mobile and web

### PR DESCRIPTION
**Pull request by khuddite**
Fixes: #6364 

1. Summary
It seems `roundingMode: 'trunc'` is not supported in the mobile runtime. Thus, the flag is simply ignored and falls back to default behavior, which seems to be a roundup

2. Solution
Replaced `roundingMode` with a manual truncation logic

3. Screenshots
CleanShot 2024-11-15 at 13 45 19
image

### CODE REVIEW

**Comment by gaearon** [comment]
We need to consider Japanese and a few other locales that don't use 3/6/9/12 as the breaking points.

**Comment by gaearon** [comment]
Here's what ChatGPT told me:

>Where Factor of 10^3 Doesn’t Work
>Here’s a summary of locales where the standard K (10^3) and M (10^6) approach doesn’t apply directly:

>- East Asia (Japan, China, Korea): Uses 10^4 (万) and 10^8 (亿, 億, 억).
>- India: Uses 10^5 (lakh) and 10^7 (crore).
>- Older European Long Scale Countries: Countries using milliard for billion (though less common today).

I think it would be good to get a large list of locales, loop over them, and then format `9`, `99`, `999`, `9999`, and so on, using our hacky polyfilled version. Then find where it doesn't look correct (i.e. where it causes `0` to show up) and fix those.

**Comment by auroursa** [comment]
Here’s a related PR: #5568. And yes, in East Asian languages, the rules for truncating numbers can differ.

**Comment by gaearon** [comment]
right, all of the PRs we've received so far on this topic don't deal with these locales — i'd take a fix that does

**Comment by khuddite** [comment]
```
const supportedLocales = [
  "af-ZA", "am-ET", "ar-AE", "ar-BH", "ar-DJ", "ar-DZ", "ar-EG", "ar-EH",
  "ar-ER", "ar-IL", "ar-IQ", "ar-JO", "ar-KM", "ar-KW", "ar-LB", "ar-LY",
  "ar-MA", "ar-MR", "ar-OM", "ar-PS", "ar-QA", "ar-SA", "ar-SD", "ar-SO",
  "ar-SS", "ar-SY", "ar-TD", "ar-TN", "ar-YE", "as-IN", "az-AZ", "be-BY",
  "bg-BG", "bn-BD", "bn-IN", "bs-BA", "ca-AD", "ca-ES", "ca-FR", "ca-IT",
  "cs-CZ", "cy-GB", "da-DK", "de-AT", "de-BE", "de-CH", "de-DE", "de-IT",
  "de-LI", "de-LU", "el-CY", "el-GR", "en-001", "en-150", "en-AE", "en-AG",
  "en-AI", "en-AS", "en-AT", "en-AU", "en-BB", "en-BE", "en-BI", "en-BM",
  "en-BS", "en-BW", "en-BZ", "en-CA", "en-CC", "en-CH", "en-CK", "en-CM",
  "en-CX", "en-CY", "en-DE", "en-DG", "en-DK", "en-DM", "en-ER", "en-FI",
  "en-FJ", "en-FK", "en-FM", "en-GB", "en-GD", "en-GG", "en-GH", "en-GI",
  "en-GM", "en-GU", "en-GY", "en-HK", "en-IE", "en-IL", "en-IM", "en-IN",
  "en-IO", "en-JE", "en-JM", "en-KE", "en-KI", "en-KN", "en-KY", "en-LC",
  "en-LR", "en-LS", "en-MG", "en-MH", "en-MO", "en-MP", "en-MS", "en-MT",
  "en-MU", "en-MW", "en-MY", "en-NA", "en-NF", "en-NG", "en-NL", "en-NR",
  "en-NU", "en-NZ", "en-PG", "en-PH", "en-PK", "en-PN", "en-PR", "en-PW",
  "en-RW", "en-SB", "en-SC", "en-SD", "en-SE", "en-SG", "en-SH", "en-SI",
  "en-SL", "en-SS", "en-SX", "en-SZ", "en-TC", "en-TK", "en-TO", "en-TT",
  "en-TV", "en-TZ", "en-UG", "en-UM", "en-US", "en-US-POSIX", "en-VC",
  "en-VG", "en-VI", "en-VU", "en-WS", "en-ZA", "en-ZM", "en-ZW", "es-419",
  "es-AR", "es-BO", "es-BR", "es-BZ", "es-CL", "es-CO", "es-CR", "es-CU",
  "es-DO", "es-EA", "es-EC", "es-ES", "es-GQ", "es-GT", "es-HN", "es-IC",
  "es-MX", "es-NI", "es-PA", "es-PE", "es-PH", "es-PR", "es-PY", "es-SV",
  "es-US", "es-UY", "es-VE", "et-EE", "eu-ES", "fa-AF", "fa-IR", "fi-FI",
  "fil-PH", "fo-FO", "fr-BE", "fr-BF", "fr-BI", "fr-BJ", "fr-BL", "fr-CA",
  "fr-CD", "fr-CF", "fr-CG", "fr-CH", "fr-CI", "fr-CM", "fr-DJ", "fr-DZ",
  "fr-FR", "fr-GA", "fr-GF", "fr-GN", "fr-GP", "fr-GQ", "fr-HT", "fr-KM",
  "fr-LU", "fr-MA", "fr-MC", "fr-MF", "fr-MG", "fr-ML", "fr-MQ", "fr-MR",
  "fr-MU", "fr-NC", "fr-NE", "fr-PF", "fr-PM", "fr-RE", "fr-RW", "fr-SC",
  "fr-SN", "fr-SY", "fr-TD", "fr-TG", "fr-TN", "fr-VU", "fr-WF", "fr-YT",
  "fy-NL", "ga-IE", "gd-GB", "gl-ES", "gu-IN", "ha-Latn-GH", "ha-Latn-NE",
  "ha-Latn-NG", "he-IL", "hi-IN", "hr-BA", "hr-HR", "hu-HU", "hy-AM",
  "id-ID", "ig-NG", "is-IS", "it-CH", "it-IT", "it-SM", "it-VA", "ja-JP",
  "ka-GE", "kk-KZ", "kl-GL", "km-KH", "kn-IN", "ko-KP", "ko-KR", "ky-KG",
  "lb-LU", "lo-LA", "lt-LT", "lv-LV", "mg-MG", "mi-NZ", "mk-MK", "ml-IN",
  "mn-MN", "mr-IN", "ms-BN", "ms-MY", "ms-SG", "mt-MT", "my-MM", "nb-NO",
  "nd-ZW", "ne-IN", "ne-NP", "nl-AW", "nl-BE", "nl-BQ", "nl-CW", "nl-NL",
  "nl-SR", "nl-SX", "nn-NO", "om-ET", "om-KE", "or-IN", "pa-Arab-PK",
  "pa-Guru-IN", "pl-PL", "ps-AF", "pt-AO", "pt-BR", "pt-CH", "pt-CV",
  "pt-GQ", "pt-GW", "pt-LU", "pt-MO", "pt-MZ", "pt-PT", "pt-ST", "pt-TL",
  "qu-BO", "qu-EC", "qu-PE", "rm-CH", "rn-BI", "ro-MD", "ro-RO", "ru-BY",
  "ru-KG", "ru-KZ", "ru-MD", "ru-RU", "ru-UA", "rw-RW", "sd-Arab-PK",
  "si-LK", "sk-SK", "sl-SI", "sn-ZW", "so-DJ", "so-ET", "so-KE", "so-SO",
  "sq-AL", "sq-MK", "sq-XK", "sr-Cyrl-BA", "sr-Cyrl-ME", "sr-Cyrl-RS",
  "sr-Cyrl-XK", "sr-Latn-BA", "sr-Latn-ME", "sr-Latn-RS", "sr-Latn-XK",
  "sv-AX", "sv-FI", "sv-SE", "sw-CD", "sw-KE", "sw-TZ", "sw-UG", "ta-IN",
  "ta-LK", "ta-MY", "ta-SG", "te-IN", "th-TH", "ti-ER", "ti-ET", "tk-TM",
  "to-TO", "tr-CY", "tr-TR", "uk-UA", "ur-IN", "ur-PK", "uz-Arab-AF",
  "uz-Cyrl-UZ", "uz-Latn-UZ", "vi-VN", "yo-BJ", "yo-NG", "zh-Hans-CN",
  "zh-Hans-HK", "zh-Hans-MO", "zh-Hans-SG", "zh-Hant-HK", "zh-Hant-MO",
  "zh-Hant-TW", "zu-ZA"
];
```

This is a list of supported locales ChatGPT gave me, I will see which locales fail with our hacky solution and make manual fixes for each of them for now.

**Comment by gaearon** [comment]
Well, don't rely on ChatGPT too much — ideally you could maybe look in the spec or something. You can also look at the `APP_LANGUAGES` constant in the app.

**Comment by khuddite** [comment]
https://playcode.io/2156494

@gaearon I know these changes might not look the nicest visually, but I've confirmed they return the same result as the previous function for numbers up to 99.9B, which I think is big enough for the followers count. I've attached playcode link that I used for testing.

File: src/view/com/util/numeric/format.ts [hunk 1]
```diff
@@ -1,12 +1,54 @@
...
+  if (i18n.locale === AppLanguage.hi) {
+    truncatedNum = truncateRounding(num, hiFactors)
+  } else if (i18n.locale === AppLanguage.ko) {
+    truncatedNum = truncateRounding(num, koFactors)
+  } else if (
+    [AppLanguage.es, AppLanguage.ca].includes(i18n.locale as AppLanguage)
+  ) {
+    truncatedNum = truncateRounding(num, esCaFactors)
+  } else if (
+    [AppLanguage.ja, AppLanguage.zh_CN, AppLanguage.zh_TW].includes(
```
Comment (auroursa) [inline src/view/com/util/numeric/format.ts:37]: I think this should also include zh-HK.

Comment (khuddite) [reply src/view/com/util/numeric/format.ts:37]: No, it doesn't apply to zh-HK

Comment (auroursa) [reply src/view/com/util/numeric/format.ts:37]: I see, thanks for your patient explanation.

**Comment by gaearon** [comment]
Fantastic work, thank you. I've incorporated your typecode as a regression test.

**Comment by khuddite** [comment]
Those updates look great. Thanks for your review!

**Review by gaearon — APPROVED** [review]

**Comment by GPHemsley** [comment]
I don't think this is the right approach, as it is conflating the lack of support for roundingMode `trunc` in the mobile runtime with the actual locale formatting of the number.

That is something that is addressed more directly (though probably also not ideally) in @mary-ext's #5568, which is attempting to fix #5538.

I don't know the React ecosystem at all, but it seems that this traces back to `@formatjs/intl-numberformat/polyfill`, which is why @pfrazee filed formatjs/formatjs#4678.

I think the better approach would be to perform the truncation of the number manually (possibly only on mobile) and then continue using the `i18n.number()` call to format it. Hardcoding specific behavior for specific locales is only going to risk making the localization process worse.

**Comment by mary-ext** [comment]
The problem with my approach is that it's still locale-dependent, not all locales have the same rounding factor.

**Comment by GPHemsley** [comment]
I was more attempting to emphasize that this approach is bypassing expertise that lives elsewhere:
* By not using `roundingMode: "trunc"` when it is available, you are losing the completely native formatting which would provide the best localized experience.
* By not calling into actual locale data in order to determine number formatting and truncation rules, you are losing the best-case fallback formatting that could also degrade the localized experience.

One more link to cross-reference here:
https://www.unicode.org/cldr/charts/46/by_type/numbers.compact_decimal_formatting.html

Note that this data captured at various implementation levels in many cases already (e.g. `Intl`, ICU). Without enumerating the reasons for needing the polyfill in the first place, it's hard to reliably point to what would be the best workaround.

**Comment by gaearon** [comment]
Can you help clarify what exact case would yield incorrect result with this approach so that we can capture it in tests? It is difficult for me to discuss abstractly as it's not my area of expertise. Having even a single incorrect example would help.

>By not using roundingMode: "trunc" when it is available, you are losing the completely native formatting which would provide the best localized experience.

Yes, it would be ideal to fall back to it — I'd need to write a feature test for it though. Or maybe it could be unconditionally used on web.

>By not calling into actual locale data in order to determine number formatting and truncation rules, you are losing the best-case fallback formatting that could also degrade the localized experience.

To clarify, AFAIK, we're only doing this for locales _supported by the app_, of which there is a limited set, and each locale supported by the app is covered by tests again `trunc` running on Node.js mode for the numbers you can see in the test case — is that insufficient to provide reasonable confidence that they match for our specific usage (one decimal point)?

Again, appreciate the concern — but a single incorrect example for supported locales would go a long way to help.

**Comment by gaearon** [comment]
For additional confidence, I've just run a `for` loop from `0` to `100000` (100K) for each locale supported by the app, and the results returned by this hacky polyfill match the results returned by the hacky polyfill in 100% cases. So this gives us decent coverage. I'm running a `for` loop up to 1 million for each of the support locales but this will take a while more.

>Note that this data captured at various implementation levels in many cases already (e.g. Intl, ICU). Without enumerating the reasons for needing the polyfill in the first place, it's hard to reliably point to what would be the best workaround.

The reason for needing the polyfill is that the Intl polyfill we use on RN does not support `trunc` rounding mode. Once that support is added upstream, we'll be able to rip this hack out.

I'll tweak the code to use the native version when it's available (on the web) in a follow-up.

**Comment by longlho** [comment]
I've fixed this in https://github.com/formatjs/formatjs/commit/50bd498a45952ad69e62d19a55b19846f33d91e9 and latest version of @formatjs/intl-numberformat polyfill should be ES2025-compliant now (pending some other ambiguity in the spec).

**Pull request merged**